langchain/src/tools/discord.ts

```diff
@@ -0,0 +1,336 @@
+    await this.client.login(this.botToken);
+
+    const guilds = await this.client.guilds.fetch();
+    await this.client.destroy();
+
+    const results =
+      guilds.map((guild) => ({
+        id: guild.id,
+        name: guild.name,
+        createdAt: guild.createdAt,
+      })) ?? [];
+
+    return JSON.stringify(results);
+  }
+}
+
+/**
+ * A tool for retrieving text channels within a server/guild a bot is a member
+ * of. It extends the base `Tool` class and implements the `_call` method to
+ * perform the retrieve operation. Requires a bot token which can be set in
+ * the environment variables. The `_call` method takes a server/guild ID
+ * to get its text channels.
+ */
+export class DiscordGetTextChannelsTool extends Tool {
+  static lc_name() {
+    return "DiscordGetTextChannelsTool";
+  }
+
+  name = "discord-get-text-channels";
+
+  description = `a discord tool. useful for getting a list of all text channels in a server/guild. input should be a discord server/guild ID`;
+
+  protected botToken: string;
+
+  client = new Client({
+    intents: [GatewayIntentBits.Guilds],
+  });
+
+  constructor(
+    botToken: string | undefined = getEnvironmentVariable("DISCORD_BOT_TOKEN")
+  ) {
+    super(...arguments);
+
+    if (!botToken) {
+      throw new Error(
+        "Discord API key not set. You can set it as DISCORD_BOT_TOKEN in your .env file."
+      );
+    }
+    this.botToken = botToken;
+  }
+
+  /** @ignore */
+  async _call(input: string): Promise<string> {
+    await this.client.login(this.botToken);
+
+    const guild = await this.client.guilds.fetch(input);
+    const channels = await guild.channels.fetch();
+    await this.client.destroy();
+
+    const results =
+      channels
+        .filter((channel) => channel?.type === ChannelType.GuildText)
+        .map((channel) => ({
+          id: channel?.id,
+          name: channel?.name,
+          createdAt: channel?.createdAt,
+        })) ?? [];
+
+    return JSON.stringify(results);
+  }
+}
+
+/**
+ * A tool for sending messages to a discord channel using a bot.
+ * It extends the base Tool class and implements the _call method to
+ * perform the retrieve operation. Requires a bot token which can be set
+ * in the environment variables. The _call method takes the message to be
+ * sent as the input argument.
+ */
+export class DiscordSendMessagesTool extends Tool {
+  static lc_name() {
+    return "DiscordSendMessagesTool";
+  }
+
+  name = "discord-send-messages";
+
+  description = `A discord tool useful for sending messages to a discod channel.
+  Input should be the discord channel message, since we will already have the channel ID.`;
+
+  protected botToken: string;
+
+  protected channelId: string;
+
+  client = new Client({
+    intents: [GatewayIntentBits.Guilds, GatewayIntentBits.GuildMessages],
+  });
+
+  constructor(
+    channelId: string,
+    botToken: string | undefined = getEnvironmentVariable("DISCORD_BOT_TOKEN")
```

Comment:
If you want to keep it as args rather than a `fields` object, let's make `botToken` always first